### DISCUSSION TITLE
关于MaaFramework和maa-node的答题器实现

### DISCUSSION THREAD

**Discussion by hanhuoer** [MaaXYZ/MaaFramework#264] [Q&A] [answered]
大佬们，最近两天开始关注这两个项目

- https://github.com/MaaXYZ/MaaFramework 
- https://github.com/neko-para/maa-node

正尝试用它实现《咸鱼之王》的答题脚本 [就是类似这个短视频的东西 https://www.bilibili.com/video/BV1ac411K7Lt](https://www.bilibili.com/video/BV1ac411K7Lt)

一开始，我是直接用最简单的 pipline 的方式试着实现，等我把几百道题目结构化后，才反应过来，感觉是不是不该这么做，可能会很慢吧，觉得任务这么多，而且是线性的。。。所以放弃

题库 pipline 片段：

![image](https://github.com/MaaXYZ/MaaFramework/assets/28647233/5ababa2e-1ff0-4f93-92f7-1ff597dff6e3)

后来，又尝试 maa-node 这个项目，试着做一点自定义动作或识别

> 此时我觉得可以以编程的方式，在中途拿到 ocr 的结果，然后去题库中迅速匹配，得到下一个需要去的指令（点击正确、错误的位置）

但刚刚调试过后才发现，好像拿不到 ocr 识别到的东西呀

---

好了，现在没有思路了。

大佬有空能看看那个视频吗，看是否适合这个场景呢？

感谢！

**Comment by moomiji** [comment, marked as answer]
编程思路

- pipeline:
``` json
{
    "MyTask": {
        "next": [
            "MyCustomRecTask"
        ]
    },
    "MyCustomRecTask": {
        "is_sub": true,
        "recognition": "OCR",
        "roi": [0, 300, 700, 300],
        "action": "Custom",
        "custom_action": "MyAct"
    }
}
```

- python
``` py
...
    maa_inst.register_action("MyAct", my_act)
...
class MyAction(CustomAction):
    def run(self, context, task_name, custom_param, box, rec_detail) -> bool:
        rec_detail          # 识别结果应该在这
        # 进行匹配
        context.click(X, Y) # 点击按钮
        return True
```

纯 pipline 的思路是用二分法，把 ocr 挪到点击按钮的 action 里：
``` json
{
    "答题_点击正确按钮": {
        "is_sub": true,
        "recognition": "OCR",
        "roi": [0, 300, 700, 300],
        "expected": [ "题一", "题二" ],
        "action": "Click",
        "target": [0, 0, 0, 0]
    },
    "答题_点击错误按钮": {
        "is_sub": true,
        "recognition": "OCR",
        "roi": [0, 300, 700, 300],
        "expected": [ "题三", "题四" ],
        "action": "Click",
        "target": [0, 0, 0, 0]
    }
}
```

**Reply by hanhuoer** [reply]
感谢回复！

回头我再试试

**Reply by hanhuoer** [reply]
没问题，就这样做就好了，感谢！
